### ISSUE TITLE
NullReferenceException in LightEpoch.cs on Dispose

### ISSUE THREAD

**Issue by resetcoder** [microsoft/FASTER#352]
I'm running a test where the same procedure is executed many times in the main thread.
In this procedure I:

1. Create **new instance of KVStore (1.7.1)** and initialize it with **one session**,
2. Try to **recover**,
3. **Add 1 million entries** with string keys, byte[] values, using SpanByte mainly as it is in the [example](https://github.com/microsoft/FASTER/blob/spanbyte/cs/samples/StoreVarLenTypes/Program.cs#L43),
4. Call **CompletePending** on the only session I have,
5. **Flush** the log,
6. **TakeFullCheckpoint**, call **CompleteCheckpointAsync** (synchronously)
7. **Iterate** through all indices,
8. **Delete the log and checkpoint directories**,
9. **Dispose** the session,
10. **Dispose** the KVStore,
11. **Dispose** the log device.
12. goto 1.

The first few (3-5) run has no problems, but after that the program crashes with `System.NullReferenceException: 'Object reference not set to an instance of an object.'` and jumps to [line #166 in **LightEpoch.cs**](https://github.com/microsoft/FASTER/blob/v1.7.1/cs/src/core/Epochs/LightEpoch.cs#L166).

Values after crash:
```
tableAligned = 0x0000000000000000
entry = 97
CurrentEpoch = 1
```

Further message:
`*tableAligned + entry = error CS0019: Operator '+' cannot be applied to operands of type 'LightEpoch.Entry' and 'int'`

_Do I mess up the Disposing sequence, or am I missing something?_ I've seen other examples where the session is stopped and the log device is closed before disposing, but as I see those methods are obsolete now and simply .Dispose should be called.

It seems (if there is) a thread/task inside FASTER doesn't finish before I'm disposing values (most probably `tableAligned`) and tries to reach an already disposed object.

**Comment by badrishc** [comment]
Do you have a call stack?

Are the Faster key and value types set to SpanByte?

Can you retry the same sample but with normal (say long) keys and values instead of SpanByte? Does it work fine then? SpanByte is new so wondering if that's contributing to the problem.

**Comment by badrishc** [comment]
Also, a repro will make this very easy to diagnose, thanks!

**Comment by badrishc** [comment]
Also, step 8 should happen after 11 (disposing log device).

**Comment by badrishc** [comment]
Another point: can you share how you are doing "Iterate through all indices"? I wonder if the iteration is not over when you call dispose.

**Comment by resetcoder** [comment]
> Do you have a call stack?

No, unfortunately I didn't get any stack info.

> Are the Faster key and value types set to SpanByte?

`FasterKV<SpanByte, SpanByte> kv;` is how I'm using it.

> Can you retry the same sample but with normal (say long) keys and values instead of SpanByte? Does it work fine then? SpanByte is new so wondering if that's contributing to the problem.

I started using FasterKV a few days ago, and gone through so many versions how it is the most useful for me, I made at least 4-5 types of key and value envelope structs, none of them worked perfectly, then I started using SpanByte and it is just working fine now for variable length types. I don't know what else should I use that doesn't just create more problems. SpanByte works fine for now if I'm "pinning" or "fixing" the memory until it finishes the copies. I don't really think SpanByte made this issue.

> can you share how you are doing "Iterate through all indices"?

```
    List<string> keys = new List<string>();
    using (IFasterScanIterator<SpanByte, SpanByte> iterator = kv.Log.Scan(kv.Log.BeginAddress, kv.Log.TailAddress))
    {
        while (iterator.GetNext(out _))
        {
            ref SpanByte keyObj = ref iterator.GetKey();
            if(!keyObj.Invalid)
                keys.Add(Encoding.ASCII.GetString(keyObj.AsSpan()));
        }
    }
    return keys;
```
The iterator had to finish, because I've used the return List<string>.Count to print how many entries are in the log, so I must had leave the using() {} block.

I'm not having the project now that caused the exception but it is not really hard to rewrite, I'm working now so I don't have time right now, but I try make a sample project that recreates the issue. The only way I could overcome the problem if I didn't dispose, just made a new instance in the same variable, therefore the GC and the finalizer finished the job, then I didn't get any exception, otherwise, always.

**Comment by resetcoder** [comment]
> Also, step 8 should happen after 11 (disposing log device).

It may be possible that caused the issue, but I'm deleting previous snapshots with active log devices, never had any issue, but it is true, I never deleted all logs. I try reproduce this later and share the code, and test the 8-11 order problem.

**Comment by badrishc** [comment]
Does it happen if you comment out just the iteration part?

**Comment by badrishc** [comment]
Could you also try the latest code in master and see if it fixes your problem as well.

**Comment by badrishc** [comment]
Btw, we will soon as `Memory<T>` support, which will make your use case much easier, no pinning etc required. See https://github.com/microsoft/FASTER/pull/354

**Comment by badrishc** [comment]
Ping on the original error report, we can close if this does not repro in latest. We tried the exact steps you outlined, bit don't see the issue. Thanks.

**Comment by badrishc** [comment]
Closing as unable to repro - request you to kindly reopen with sample code, if you are still experiencing the issue.

**Comment by resetcoder** [comment]
Sorry for the delay, I've just got some time to recreate the project.

While recreating I actually could reproduce the issue, which is good, but I also noticed I've used a wrong disposing scheme, and with the proper scheme it is working fine, so it is my bad, I'm sorry.

I upload both projects anyways, but the problem was running the finalizer and the object disposals in the wrong order.

After fixing the disposing scheme the test could run to the end without any crashes.

Sorry again for my mistake, but if You want to look at it, then here are the projects:

[FASTER_Test_Wrong.zip](https://github.com/microsoft/FASTER/files/5405586/FASTER_Test_Wrong.zip)
[FASTER_Test_Proper.zip](https://github.com/microsoft/FASTER/files/5405588/FASTER_Test_Proper.zip)

Btw is there any ways to contact You besides this issues page? I'd have a few questions about using the log/compaction and I don't think those question belong here. Thank You in advance!
